mainwindow.cpp

```diff
@@ -0,0 +1,334 @@
+    MailHistoryUnit* new_history_unit = new MailHistoryUnit(Letters);
+
+    if (layout)
+    {
+        new_history_unit = new MailHistoryUnit(Letters);
+        layout->setAlignment(Qt::AlignHCenter | Qt::AlignTop);
+        layout->insertWidget(0, new_history_unit);
+    }
+
+    connect(new_history_unit, SIGNAL(OnMouseReleased(QVector<LetterStruct>)), this, SLOT(HistoryWidgetClicked(QVector<LetterStruct>)));
+}
+
+void MainWindow::PopulateMailsHistory()
+{
+    QDir dir(m_temp_file_path);
+
+    if (!dir.exists())
+    {
+        qDebug() << "Directory does not exist:" << m_temp_file_path;
+    }
+
+    dir.setSorting(QDir::Time);
+    dir.setFilter(QDir::Files);
+    QStringList files = dir.entryList();
+
+    for (const QString &fileName : files)
+    {
+        qDebug() << fileName;
+        QVector<LetterStruct> letters = ReadLettersFromFile(m_temp_file_path + fileName);
+        if (!letters.isEmpty()) SpawnNewHistoryUnit(letters);
+    }
+}
+
+void MainWindow::SendLetter(const LetterStruct& Letter)
+{
+
+}
+
+bool MainWindow::WriteLettersToFile(const QVector<LetterStruct>& letters, const QString& full_file_name)
+{
+    QFile file(full_file_name);
+    if (!file.open(QIODevice::ReadWrite))
+    {
+        qDebug() << "Could not open file for writing";
+        return false;
+    }
+    if (letters.isEmpty())
+    {
+        qDebug() << "Trying to write 0 letters!";
+        return false;
+    }
+
+    qDebug() << "Writing to " << full_file_name << " " << letters.size() << " letters";
+
+    QDataStream out(&file);
+    out.setVersion(QDataStream::Qt_6_7);
+
+    out << letters.size();
+
+    for (const auto& letter : letters)
+    {
+        out << letter;
+    }
+
+    file.close();
+
+    return true;
+}
+
+QVector<LetterStruct> MainWindow::ReadLettersFromFile(const QString& full_file_name)
+{
+    QFile file(full_file_name);
+    QVector<LetterStruct> Letters;
+
+    if (!file.open(QIODevice::ReadOnly))
+    {
+        qDebug() << "Could not open file " << full_file_name << " for reading";
+        return Letters;
+    }
+    if (!QFile::exists(full_file_name))
+    {
+        qDebug() << "The file " << full_file_name << " was not found";
+        return Letters;
+    }
+
+    qDebug() << "Reading from: " << full_file_name;
+
+    QDataStream in(&file);
+
+    qsizetype size;
+    in >> size;
+    qDebug() << size;
+
+    for (qsizetype i = 0; i < size; i++)
+    {
+        LetterStruct Letter;
+        in >> Letter;
+        Letters.append(Letter);
+    }
+
+    file.close();
+    return Letters;
+}
+
+bool MainWindow::AppendLettersToFile(const QVector<LetterStruct>& Letters, const QString& FullFileName)
+{
+    QVector<LetterStruct> existing_letters = ReadLettersFromFile(FullFileName);
+
+    existing_letters.append(Letters);
+
+    WriteLettersToFile(existing_letters, FullFileName);
+
+    return true;
+}
+
+void MainWindow::SelectFilesAndRefreshLabels()
```

Comment:
functions have to do only one job. labels refreshing could be moved in a separate function

Reply:
But it makes no sense to select files and not refresh labels, does it? I agree that "functions have to do only one job", however these two jobs are always together. It would make no sense to separate them, wouldn't it?